feat: 게시글 좋아요 기능 구현

dhfkdlsj opened the pull request:
## 개요
- 게시글 좋아요 / 좋아요 취소 기능 구현
- Redis에 좋아요 상태 저장
- 주기적으로 Postgre DB와 동기화
- 게시글 상세조회시 해당 게시글의 좋아요 관련 정보 캐싱
- 테스트코드 수정

Resolves: #31(게시글 좋아요 기능 구현)

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

dhfkdlsj committed:
Feat: 좋아요 기능 추가

- 좋아요를 누르면 Redis -> Postgre 순서로 확인
- 관련 데이터를 Redis에 캐싱
- 사용자의 행동을 로그로 남겨서 Redis에 저장

Resolves: #31

dhfkdlsj committed:
Feat: 게시글 상세 조회 시 좋아요 관련 정보 제공

- 게시글 상세 조회 시 해당 게시글에 대한 사용자의 좋아요 여부 정보 제공
- 해당 게시글의 좋아요 수 정보 제공
- 해당 데이터들은 Redis를 통해 조회

Resolves: #31

dhfkdlsj committed:
Feat: 주기적으로 Redis의 로그를 읽어 Postgre 데이터베이스의 데이터 갱신

- Test를 위해 5초 간격으로 실행
- Redis의 로그를 읽어서, Postgre 데이터베이스의 PostLike 테이블에 데이터 추가 또는 삭제
- Post 테이블의 LikeCount 컬럼 값 갱신

Resolves: #31

dhfkdlsj committed:
Feat: 스케쥴러 사용을 위한 어노테이션 추가

- 스케쥴러 사용을 위해 @EnableScheduling 어노테이션 추가

Resolves: #31

dhfkdlsj committed:
Fix: 병합 시 충돌 해결

- 병합 시 충돌 해결

Resolves: #31

dhfkdlsj committed:
Feat: 게시글 좋아요 시 유효한 게시글인지 검증

- 해당 게시글이 존재하는지 확인
- SoftDelete된 게시글인지 확인

Resolves: #31

dhfkdlsj committed:
Refactor: 좋아요 로그 남기는 방식을 Redis Stream -> Redis Hash로 변경

- 로그를 스트림 한 번에 처리하는 것에서 해시로 변경
- 중복 로그를 방지하고 하나씩 처리함으로써 동시성제어

Resolves: #31

dhfkdlsj committed:
Refactor: 게시글에 좋아요를 누를때도 룩 어사이드 패턴을 사용하도록 코드 수정

- 게시글에 좋아요를 눌렀을때도 레디스에 관련 데이터 확인 후 없으면 디비에 다녀오도록 수정

Resolves: #31

dhfkdlsj committed:
Test: 서비스 코드 리팩토링과 PostLike 추가에 따른 테스트 코드 변경

- 서비스 코드 리팩토링과 PostLike 추가에 따른 테스트 코드 변경

Resolves: #31

heodongwon commented on src/main/java/com/itcen/censquare/domain/postlike/repository/PostLikeRepository.java:
```diff
@@ -0,0 +1,26 @@
+  void deleteBy(@Param("postId") Long postId, @Param("memberId") Long memberId);
+
+  @Query("SELECT COUNT(pl) > 0 FROM PostLike pl WHERE pl.post.postId = :postId AND pl.member.memberId = :memberId")
+  boolean existsBy(@Param("postId") Long postId, @Param("memberId") Long memberId);
```
existsBy가 동일한 기능을 하는 메서드로 보이는데 굳이 따로 두신 이유가 있을까요?

dhfkdlsj replied:
넵! 
자동 파싱 메서드(existsByPostPostIdAndMemberMemberId)와 동일한 기능을 하는 것을 미처 인지하지 못했네요. 
말씀 주신 대로 수동 쿼리 메서드는 제거하고 정리해서 반영하겠습니다 :) 
빠르게 수정하겠습니다~! 감사합니다! 🙇‍♂️

HaegyeongKim01 commented on src/main/java/com/itcen/censquare/domain/postlike/service/RedisLikeService.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  public boolean isPostLikedByUser(Long postId, Long memberId) {
+    String key = "post:like:set:" + postId;
```
"post:like:set:" 이 부분은 상단에 정의된 `LIKE_SET_PREFIX`를 사용해도 좋을 것 같아요!  :)

dhfkdlsj replied:
감사합니다 바로 수정하겠습니다~!!

heodongwon commented on src/main/java/com/itcen/censquare/domain/postlike/service/RedisLikeService.java:
```diff
@@ -0,0 +1,99 @@
+    return count;
+  }
+
+  public boolean isPostLikedByUser(Long postId, Long memberId) {
```
Transactional의 readOnly 빠진 것 같습니다

dhfkdlsj replied:
해당 메서드는 아래와 같이 조건부로 Redis에 캐싱하는 쓰기 동작이 포함되어 있어 readOnly = true는 오히려 적절하지 않다고 판단했는데 어떻게 생각하시나요??
`redisTemplate.opsForSet().add(key, memberId.toString());`

dhfkdlsj committed:
Refactor: 중복 코드 제거

- existsByPostIdAndMemberId가 중복되던 부분 제거

Resolves: #31

dhfkdlsj committed:
Refactor: 게시글 좋아요 Redis 키 prefix 상수화

- 게시글 좋아요 Redis 키 prefix 상수화

Resolves: #31

dhfkdlsj committed:
Refactor: getLikeCount 메서드에 읽기전용 속성 해제

- getLikeCount 메서드에 읽기전용 속성 해제

Resolves: #31

heodongwon reviewed (approved):
고생하셨습니다!

hj-k66 commented on src/main/java/com/itcen/censquare/domain/postlike/service/RedisLikeService.java:
```diff
@@ -0,0 +1,98 @@
+      redisTemplate.opsForSet().remove(key, value);
+      log.info("💡 Redis 캐시 미스, DB에는 존재 → UNLIKE 처리");
+      return false;
+    }
```
왜 다시 캐싱하고 삭제하는지 알 수 있을까요?

dhfkdlsj replied:
저 부분은 레디스에 데이터가 없고, 디비에 존재할때 명시적으로 보여주기위해서 적어둔건데, 굳이 필요 없는 코드 같습니다!
그리고 현재는 로그를 따로 관리해서, 로그 바로 저장하는 방식으로 수정해보겠습니다!~!
감사합니다!!!!

dhfkdlsj committed:
Refactor: Redis에 좋아요 정보 캐싱하는 방식 변경

- 해당 게시글에 좋아요를 누른 모든 사용자 리스트를 저장하던 방식에서 좋아요 수를 저장하는 방식으로 변경
- Lua 스크립트를 사용하여 동시성제어

Resolves: #31

dhfkdlsj committed:
Test: 좋아요 기능 동시성 제어 테스트 코드 작성

- 좋아요 기능 동시성 제어 테스트 코드 작성
- 한 사용자가 100번 눌렀을 경우와 100명의 사용자가 1번씩 누르고, 50명의 사용자는 1번씩 더 누른 경우 테스트

Resolves: #31

dhfkdlsj committed:
Fix: 병합 시 충돌 해결

- 병합 시 충돌 해결

Resolves: #31

hj-k66 reviewed (approved):
굿👍
